### PR TITLE
Merging parallel format parsing

### PR DESCRIPTION
**Pull request by akuzm**
Merging #6553

Changelog category (leave one):
- New Feature

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):
Speed up parsing of data files in CSV, TSV and JSONEachRow format by doing it in parallel. This is enabled by default and controlled by the `max_threads_for_parallel_parsing` setting.

TODO:
* Fall back to parsing without segmentation in a single-threaded case -- the single-threaded regression is noticeable in Kafka tests.
* rewrite eofWithSavingWhatever.... to something sane. 
* fix a mysterious bug:
```
SELECT UserID, SearchPhrase, count() FROM table GROUP BY UserID, SearchPhrase limit 1 settings max_threads_for_parallel_parsing = 1
will use 1 threads

SELECT UserID, SearchPhrase, count() FROM table GROUP BY UserID, SearchPhrase limit 1 format PrettySpace settings max_threads_for_parallel_parsing = 1
will use 6 threads
```

DONE:
* Remove the MemoryExt<> wrapper that is not needed after this fix: #7797 
* Move InputStream creation from segmentator to parser or avoid it altogether -- it's noticeable in perf reports.

### CODE REVIEW

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.h [hunk 1]
```diff
@@ -193,8 +193,10 @@ class ParallelParsingBlockInputStream : public IBlockInputStream
         pool.scheduleOrThrowOnError(std::bind(&ParallelParsingBlockInputStream::parserThreadFunction, this, unit_number));
     }
 
-    void waitForAllThreads()
+    void finishAndWait()
     {
+        finished.store(true, std::memory_order_release);
```
Comment (nikitamikhaylov) [inline dbms/src/DataStreams/ParallelParsingBlockInputStream.h]: Why do you use *std::memory_order_release* there and don't use *std::memory_order_acquire* somewhere else? Also I see *finished = true* in other part of code where default sequential consistency model is used.

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp [hunk 2]
```diff
@@ -106,52 +101,100 @@ void ParallelParsingBlockInputStream::parserThreadFunction(size_t current_unit_n
...
+        next_block_in_current_unit = 0;
+    }
+
+    assert(next_block_in_current_unit);
+    assert(!unit.is_past_the_end);
+
+    Block res = std::move(unit.block_ext.block.at(*next_block_in_current_unit));
+    last_block_missing_values = std::move(unit.block_ext.block_missing_values[*next_block_in_current_unit]);
+
+    ++*next_block_in_current_unit;
```
Comment (nikitamikhaylov) [inline dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp]: OMG. [Maybe this is better?](https://en.cppreference.com/w/cpp/utility/optional/value)

File: dbms/src/Core/Settings.h [hunk 3]
```diff
@@ -108,6 +108,10 @@ struct Settings : public SettingsCollection<Settings>
     M(SettingBool, distributed_group_by_no_merge, false, "Do not merge aggregation states from different servers for distributed query processing - in case it is for certain that there are different keys on different shards.") \
     M(SettingBool, optimize_skip_unused_shards, false, "Assumes that data is distributed by sharding_key. Optimization to skip unused shards if SELECT query filters by sharding_key.") \
     \
+    M(SettingBool, input_format_parallel_parsing, true, "Enable parallel parsing for several data formats (JSONEachRow, TSV, TKSV, CSV).") \
+    M(SettingUInt64, max_threads_for_parallel_parsing, 10, "The maximum number of threads to parallel parsing.") \
+    M(SettingUInt64, min_chunk_size_for_parallel_parsing, (1024 * 1024), "The minimum chunk size in bytes, which each thread will parse in parallel.") \
```
Comment (akuzm) [inline dbms/src/Core/Settings.h]: Should be min_chunk_bytes, not size.

Comment (nikitamikhaylov) [reply dbms/src/Core/Settings.h]: > The minimum chunk size in bytes.

**Comment by akuzm** [comment]
The main reason for the ugliness and slowness in the segmentation engine functions is that the ReadBuffer does not allow to read to the externally supplied memory. For now I am emulating this with memcpys, which show up in profiles as the major thing the segmentation engine does. We need some other API that allows to do a system read() directly to the externally supplied memory.

This could be improved later, but the current code (eofWithSaving... and friends) looks too convoluted to merge as is.

**Comment by nikitamikhaylov** [comment]
But we can disable this feature by default or mark it as experimental.

**Comment by akuzm** [comment]
There is `ReadBuffer::readBig` that may be suitable, but it's not implemented for plain files, only for compressed ones. Should be easy to do.

**Comment by akuzm** [comment]
> But we can disable this feature by default or mark it as experimental.

I think I'll start with merging the segmentation for TSV that I already rewrote, and then we can add other formats one by one.
For now, I'll have to rewrite TSV segmentation once again so that it doesn't do so much copying :see_no_evil:

**Comment by akuzm** [comment]
> There is `ReadBuffer::readBig` that may be suitable, but it's not implemented for plain files, only for compressed ones. Should be easy to do.

Turned out to be not easy at all -- being basically an unbuffered read (`ReadBuffer::readUnbuffered`..?) , it has unclear interplay with `nextImpl()`. This probably requires architectural changes. But as an experiment, I did manage to make it work for a particular case, giving a max of 1.57GB/s TSV parsing speed on 12 threads (6 cores w/hyperthreading), and 1.30 on 6 threads. The segmentator thread spends almost all time in system calls that way.

Thinking some more, the most easy and most performant way for a normal file would probably be to mmap() it, but we still have to support sequential reads for pipes etc.

**Comment by akuzm** [comment]
```
┌─threads─┬─median time─┬─GB/s─┬─per thread─┐
│       1 │       27.29 │ 0.28 │       0.28 │
│       2 │       14.58 │ 0.54 │       0.27 │
│       3 │       10.17 │ 0.77 │       0.25 │
│       4 │        8.30 │ 0.94 │       0.23 │
│       5 │        7.09 │ 1.11 │       0.22 │
│       6 │        6.77 │ 1.16 │       0.19 │
│       7 │        6.25 │ 1.26 │       0.18 │
│       8 │        6.18 │ 1.27 │       0.15 │
│       9 │        6.05 │ 1.30 │       0.14 │
│      10 │        5.97 │ 1.31 │       0.13 │
│      11 │        5.86 │ 1.34 │       0.12 │
│      12 │        5.89 │ 1.33 │       0.11 │
│      13 │        5.87 │ 1.34 │       0.10 │
│      14 │        5.98 │ 1.31 │       0.09 │
└─────────┴─────────────┴──────┴────────────┘
```
The parsing throughput doesn't scale linearly with the number of threads, not sure why. I'm parsing hits_v1.tsv on 6 cores with HT.

**Comment by nikitamikhaylov** [comment]
image
Speedup scales almost linearly up to 6 threads.

**Comment by akuzm** [comment]
```
┌────chunk─┬─median time─┐
│     1000 │      294.10 │
│    10000 │       54.37 │
│   100000 │       14.14 │
│  1000000 │        9.47 │
│ 10000000 │        8.61 │
└──────────┴─────────────┘
```

10MB chunk is even faster than 1MB
(hits_v1.tsv, 6 threads total)

File: docs/en/operations/settings/settings.md [hunk 4]
```diff
@@ -994,4 +994,20 @@ Lower values mean higher priority. Threads with low `nice` priority values are e
...
 Default value: 0.
 
+## input_format_parallel_parsing
+
+- Type: bool
+- Default value: True
+
+Enable order-preserving parallel parsing of data formats such as JSONEachRow, TSV, TKSV and CSV. Reading will be single threaded and parsing will be multithreaded.
+
+The maximum number of threads for order-preserving parallel parsing of data formats. 0 means use global maximum.
```
Comment (akuzm) [inline docs/en/operations/settings/settings.md]: Will push the fix for this after the tests run.

**Pull request merged**